content/en/docs/advanced-tasks/unsafe-fail-forward-upgrades.md

```diff
@@ -0,0 +1,123 @@
+
+By opting into "failed forward" upgrades, OLM will allow you to recover from failed installations and upgrades by:
+- Allowing CSVs to move from the FAILED phase to the REPLACING phase when "UnsafeFailForward" upgrades are enabled.
+- Allowing OLM to calculate new installPlans for a set of installables if:
+  - The InstallPlan referenced by a subscription is in the FAILED phase.
+  - The CatalogSource has been updated to include a new upgrade for one or more CSVs in the namespace.
+
+## Using UnsafeFailForward Upgrades
+
+Since resolution is namespace scoped, the toggle for allowing "UnsafeFailForward" upgrades is namespace scoped as well. Accordingly, the namespace scoped resource, OperatorGroup, has the `upgradeStrategy` field.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: foo
+  namespace: bar
+spec:
+  upgradeStrategy:
+     # Possible values include "Default" or "TechPreviewUnsafeFailForward".
+    name: TechPreviewUnsafeFailForward
+```
+
+With the upgradeStrategy type set to `TechPreviewUnsafeFailForward`, OLM will allow operators to "UnsafeFailForward" in adherence with the principles discussed below. If the upgradeStrategy is unset or set to `Default` OLM will exhibit existing behavior.
+
+## Understanding "UnsafeFailForward" upgrades
+
+Before using UnsafeFailForward upgrades, it is important to understand how OLM and the Resolver calculate what needs to be installed. When determining what to install, OLM provides the resolver with the set of CSVs in a namespace. The resolver treats these CSVs differently depending on whether or not they are claimed by a Subscription, where the Subscription lists the CSV's name in its `.status.currentCSV` field.
+- If a CSV is claimed by a Subscription, the resolver will allow it to be upgraded by a bundle that replaces it.
+- If a CSV is not claimed by a Subscription, it is assumed that the user installed the CSV themselves, that OLM should not upgrade the CSV, and that it must appear in the next set of installables.
+
+OLM's Resolver is deterministic, meaning that the set of installables will not change for a given set of arguments if no new upgrades have been declared in the existing CatalogSources. When an upgrade fails due to a CSV entering the FAILED phase, the CSV being replaced still exists and is not referenced by a subscription. Today, OLM would send both `Operator v1` and `Operator v2` to the resolver which will normally be unsatisfiable because `Operator v1` is marked as required and `Operator v2` cannot be upgraded further since it provides the same APIs as `Operator v1`. In order to support "failing forward", OLM needs to omit CSVs in the REPLACING phase from the set of arguments sent to the resolver when "UnsafeFailForward" upgrades are enabled.
```

Comment:
```suggestion
OLM's Resolver is deterministic, meaning that the set of installables will not change for a given set of arguments if no new upgrades have been declared in the existing CatalogSources. When an upgrade fails due to a CSV entering the FAILED phase, the CSV being replaced still exists and is not referenced by a Subscription. Today, OLM would send both `Operator v1` and `Operator v2` to the Resolver which will normally be unsatisfiable because `Operator v1` is marked as required and `Operator v2` cannot be upgraded further since it provides the same APIs as `Operator v1`. In order to support "failing forward", OLM needs to omit CSVs in the REPLACING phase from the set of arguments sent to the Resolver when "UnsafeFailForward" upgrades are enabled.
```